### PR TITLE
Typescript migration

### PR DESCRIPTION
**Pull request by djhouseknecht**
Review and merge https://github.com/MyPureCloud/purecloud-webrtc-sdk/pull/24 first.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -78,20 +63,33 @@
     "wildemitter": "^1.2.0"
   },
   "devDependencies": {
-    "ava": "^0.20.0",
+    "@types/jest": "^24.0.15",
+    "@types/nock": "^10.0.3",
+    "@types/safe-json-stringify": "^1.1.0",
+    "@types/sinon": "^7.0.13",
```
Comment (ckbrewer33) [inline package.json]: Jest has it's own mocking library built into it, so you shouldn't still need sinon

File: package.json [hunk 2]
```diff
@@ -78,20 +63,33 @@
...
     "babelify": "^8.0.0",
     "browserify": "^16.2.3",
     "gulp": "^4.0.0",
     "gulp-uglify": "^3.0.1",
     "gulp-util": "^3.0.8",
+    "jest": "^24.8.0",
     "nock": "^9.3.0",
-    "nyc": "^11.1.0",
     "semistandard": "^12.0.1",
     "sinon": "^7.1.1",
```
Comment (ckbrewer33) [inline package.json]: Same as above

File: test/test-utils.ts [hunk 3]
```diff
@@ -0,0 +1,225 @@
+import WebSocket from 'ws';
+import nock from 'nock';
+// import sinon from 'sinon';
```
Comment (ckbrewer33) [inline test/test-utils.ts]: Don't need this comment

File: test/unit/logging.test.ts [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+      wss.removeAllListeners();
+    }
+    sandbox.restore();
+  });
+
+  test('_log | will not notify logs if the logLevel is lower than configured', async () => {
+    const { sdk } = mockApis({ withLogs: true });
+    sdk._logLevel = 'warn';
+    await sdk.initialize();
+    sinon.stub(sdk._backoff, 'backoff'); // called in notifyLogs
```
Comment (ckbrewer33) [inline test/unit/logging.test.ts]: You can use jest's mocking library instead here, and eliminate sinon entirely.
`jest.spyOn(sdk._backoff, 'backoff')`

File: test/unit/logging.test.ts [hunk 5]
```diff
@@ -0,0 +1,323 @@
...
+    log.call(sdk, 'debug', 'test', { details: 'etc' });
+    await timeout(1100);
+    sinon.assert.notCalled(sdk._backoff.backoff);
```
Comment (ckbrewer33) [inline test/unit/logging.test.ts]: `expect(sdk._backoff.backoff).not.toHaveBeenCalled()`

File: test/unit/logging.test.ts [hunk 6]
```diff
@@ -0,0 +1,323 @@
...
+    expect(sdk._reduceLogPayload).toBe(false);
+    sdk._logBuffer = [];
+  });
+
+  test('_sendLogs | resets the backoff on success', async () => {
+    const { sdk } = mockApis({ withLogs: true });
+    sdk._logLevel = 'warn';
+    await sdk.initialize();
+
+    const backoffResetSpy = sinon.spy(sdk._backoff, 'reset');
```
Comment (ckbrewer33) [inline test/unit/logging.test.ts]: `jest.spyOn(sdk._backoff, 'reset');`

File: test/unit/logging.test.ts [hunk 7]
```diff
@@ -0,0 +1,323 @@
...
+    sdk._logBuffer.push('log1');
+    sdk._logBuffer.push('log2');
+
+    sdk._backoff.backoff();
+    await timeout(100);
+    sinon.assert.calledOnce(backoffResetSpy);
```
Comment (ckbrewer33) [inline test/unit/logging.test.ts]: `expect(sdk._backoff.reset).toHaveBeenCalledOnce()`

**Comment by djhouseknecht** [comment]
Some of the types could be stronger; there are still some implicit any types. We can work on slowly changing those over as we keep doing work on the repo. I wanted to get the initial migration out the door so we could start working on the screen-share portion.

File: package.json (hunk 1 above)
Comment (djhouseknecht) [reply package.json]: Yeah, you are right. Initially I just wanted to get the tests working with as little changes as possible. I would like to remove `sinon` for sure. Maybe I will start doing that now that the main portion of the migration is complete.

**Pull request merged**